### ISSUE TITLE
[Bug]: flatpak update --no-deploy fails when the operation would cause an EOL runtime to be removed

### ISSUE THREAD

**Issue by wjt** [flatpak/flatpak#5171]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.14.0

### What Linux distribution are you using?

Fedora Linux

### Linux distribution version

37

### What architecture are you using?

x86_64

### How to reproduce

1. Ensure that you have exactly one app installed which uses a runtime marked end-of-life, with an available update to a non-EOL runtime.

In my case I encountered this with a version of `org.laptop.TurtleArtActivity` that uses `org.gnome.Platform//3.34`, and you can recreate this situation as follows:

```bash
flatpak install org.laptop.TurtleArtActivity
sudo flatpak update --commit=f399bc4a0562b7cf17791cb8b54c1f1620e79e160d1b8614d44ef26d7870dc4b org.laptop.TurtleArtActivity
```

2. Run `flatpak update --no-deploy`

### Expected Behavior

`--no-deploy` is documented as follows:

> Download the latest version, but don't deploy it.

It exists to be subsequently paired with `flatpak update --no-pull`, to install whatever updated versions are locally available.

So I would expect that running this command would download all available updates, and any new runtimes that become needed by the app would be downloaded, but nothing would be installed, upgraded, or removed.

### Actual Behavior

The operation removes the Locale and other extensions of the 3.34 runtime, and then tries to remove `org.gnome.Platform//3.34` itself. This fails, because the `org.laptop.TurtleArtActivity` update that would cause this EOL runtime to become unused has not actually been deployed, only pulled. Here is the full output:

```
$ flatpak update --no-deploy 
Looking for updates…

Info: runtime org.gnome.Sdk branch 41 is end-of-life, with reason:
   The GNOME 41 runtime is no longer supported as of September 17, 2022. Please ask your application developer to migrate to a supported platform.

Info: runtime org.gnome.Platform branch 3.34 is end-of-life, with reason:
   The GNOME 3.34 runtime is no longer supported as of 14th August 2020. Please ask your application developer to migrate to a supported platform.
Info: applications using this runtime:
   org.laptop.TurtleArtActivity

Info: runtime org.freedesktop.Platform.VAAPI.Intel branch 19.08 is end-of-life, with reason:
   The Freedesktop SDK 19.08 runtime is no longer supported as of September 1, 2021. Please ask your application developer to migrate to a supported version
Info: applications using this extension:
   org.laptop.TurtleArtActivity

Info: runtime org.freedesktop.Platform.GL.default branch 19.08 is end-of-life, with reason:
   The Freedesktop SDK 19.08 runtime is no longer supported as of September 1, 2021. Please ask your application developer to migrate to a supported version
Info: applications using this extension:
   org.laptop.TurtleArtActivity


        ID                                                  Branch           Op          Remote           Download
 1. [✓] com.transmissionbt.Transmission.Locale              stable           u           flathub            1.0 kB / 962.7 kB
 2. [✓] org.ardour.Ardour.Locale                            stable           u           flathub            1.0 kB / 6.8 MB
 3. [✓] org.freedesktop.Platform.Compat.i386                22.08            u           flathub            1.1 kB / 106.2 MB
 4. [✓] org.freedesktop.Platform.GL.default                 22.08            u           flathub            1.0 kB / 131.0 MB
 5. [✓] org.freedesktop.Platform.GL32.default               22.08            u           flathub            1.0 kB / 138.4 MB
 6. [✓] org.freedesktop.Platform.Locale                     22.08            u           flathub            1.0 kB / 333.0 MB
 7. [✓] org.freedesktop.Platform.VAAPI.Intel.i386           22.08            u           flathub            1.0 kB / 1.9 MB
 8. [✓] org.freedesktop.Platform.VAAPI.Intel                22.08            u           flathub            1.0 kB / 10.8 MB
 9. [✓] org.freedesktop.Platform.ffmpeg-full                22.08            u           flathub            1.0 kB / 4.3 MB
10. [✓] org.freedesktop.Platform                            22.08            u           flathub            1.1 kB / 214.4 MB
11. [✓] org.signal.Signal                                   stable           u           flathub            1.0 kB / 171.7 MB
12. [✓] org.ardour.Ardour                                   stable           u           flathub            1.0 kB / 45.6 MB
13. [✓] im.riot.Riot                                        stable           u           flathub            1.0 kB / 111.8 MB
14. [✓] org.freedesktop.Sdk.Locale                          22.08            u           flathub            1.0 kB / 338.7 MB
15. [✓] org.freedesktop.Sdk                                 22.08            u           flathub            1.1 kB / 512.0 MB
16. [✓] org.vim.Vim                                         stable           u           flathub            1.0 kB / 14.2 MB
17. [✓] org.gnome.Builder.Locale                            stable           u           flathub            1.0 kB / 3.9 MB
18. [✓] com.transmissionbt.Transmission                     stable           u           flathub            1.0 kB / 1.5 MB
19. [✓] org.gnome.Platform.Locale                           43               u           flathub            1.0 kB / 340.2 MB
20. [✓] org.gnome.Platform                                  43               u           flathub            1.1 kB / 325.0 MB
21. [✓] com.dropbox.Client                                  stable           u           flathub          113.7 MB / 114.2 MB
22. [✓] org.gnome.Sdk.Locale                                43               u           flathub            1.0 kB / 344.9 MB
23. [✓] org.gnome.Sdk                                       43               u           flathub            1.1 kB / 658.5 MB
24. [✓] org.gnome.Builder                                   stable           u           flathub            1.0 kB / 195.7 MB
25. [✓] com.borgbase.Vorta                                  stable           u           flathub            1.0 kB / 36.1 MB
26. [✓] org.kde.ktuberling.Locale                           stable           u           flathub            1.0 kB / 309.9 kB
27. [✓] org.kde.ktuberling                                  stable           u           flathub            1.0 kB / 63.5 MB
28. [✓] org.laptop.TurtleArtActivity                        stable           u           flathub            1.0 kB / 13.2 MB
29. [✓] org.mozilla.firefox.Locale                          stable           u           flathub            1.0 kB / 48.1 MB
30. [✓] org.mozilla.firefox                                 stable           u           flathub            1.0 kB / 89.1 MB
31. [-] org.freedesktop.Platform.GL.default                 19.08            r
32. [-] org.freedesktop.Platform.VAAPI.Intel                19.08            r
33. [-] org.gnome.Platform.Locale                           3.34             r
34. [✗] org.gnome.Platform                                  3.34             r

Error: Can't remove org.gnome.Platform/x86_64/3.34, it is needed for: org.laptop.TurtleArtActivity
Changes complete.
error: There were one or more errors
$ echo $?
1
```

### Additional Information

The [same bug has been reported against GNOME Software](https://gitlab.gnome.org/GNOME/gnome-software/-/issues/1711), which as part of its normal operation performs (the libflatpak API equivalent of) a `flatpak update --no-deploy` followed by `flatpak update --no-pull`. Personally [I think it shouldn't](https://gitlab.gnome.org/GNOME/gnome-software/-/issues/1929) but nonetheless the current behaviour of `flatpak update --no-deploy` is broken in this situation.

I think that a transaction that has had `flatpak_transaction_set_no_deploy (t, TRUE)` set should either not add operations to uninstall unused runtimes, or uninstall operations should be a no-op if no_deploy is set.

**Comment by wjt** [comment]
I am about to find a similar bug about the behaviour of eol-rebase'd apps with `--no-deploy`, which leads me to think that making uninstall ops do nothing if `no_deploy` is set might be the right fix.

**Comment by dbnicholson** [comment]
I think this would also be fixed by #5289, which doesn't queue any uninstall operations if the transaction is set to not deploy.

**Comment by dbnicholson** [comment]
I believe this sold also be closed by #5289. I forgot to include the Fixes for this in the commit message.
